File: kotlin/Sort and Search/SortArrayCountingSort.kt
fun sortArrayCountingSort(nums: List<Int>): List<Int> {
    if (nums.isEmpty()) return emptyList()
    val res = mutableListOf<Int>()
    // Count occurrences of each element in 'nums'.
    val counts = IntArray(nums.maxOrNull()!! + 1)
    for (num in nums) {
        counts[num]++
    }
    // Build the sorted array by appending each index 'i' to it a total
    // of 'counts[i]' times.
    for (i in counts.indices) {
        val count = counts[i]
        repeat(count) { res.add(i) }
    }
    // counts.forEachIndexed { i, count -> repeat(count) { res.add(i) } }
    return res
}

File: cpp/Sort and Search/sort_array_counting_sort.cpp
#include <vector>
#include <algorithm>

std::vector<int> sortArrayCountingSort(std::vector<int>& nums) {
    if (nums.empty()) {
        return {};
    }
    std::vector<int> res;
    // Count occurrences of each element in 'nums'.
    int maxNum = *std::max_element(nums.begin(), nums.end());
    std::vector<int> counts(maxNum + 1, 0);
    for (int num : nums) {
        counts[num]++;
    }
    // Build the sorted array by appending each index 'i' to it a total
    // of 'counts[i]' times.
    for (int i = 0; i <= maxNum; i++) {
        for (int j = 0; j < counts[i]; j++) {
            res.push_back(i);
        }
    }
    return res;
}